bin/host/src/orchestrator.rs

```diff
@@ -0,0 +1,132 @@
+//! Contains the [HostOrchestrator] trait, which defines entry points for the host to run a given
+//! module.
+
+use crate::{Fetcher, PreimageServer, SharedKeyValueStore};
+use anyhow::Result;
+use async_trait::async_trait;
+use kona_preimage::{
+    BidirectionalChannel, HintReader, HintWriter, NativeChannel, OracleReader, OracleServer,
+};
+use kona_std_fpvm::{FileChannel, FileDescriptor};
+use std::sync::Arc;
+use tokio::{sync::RwLock, task};
+
+/// The host<->client communication channels. The client channels are optional, as the client may
+/// not be running in the same process as the host.
+#[derive(Debug)]
+struct HostComms {
+    /// The host<->client hint channel.
+    pub hint: BidirectionalChannel,
+    /// The host<->client preimage channel.
+    pub preimage: BidirectionalChannel,
+}
+
+/// The host->client communication channels when running in detached mode. The client channels are
+/// held in a separate process.
+#[derive(Debug)]
+struct DetachedHostComms {
+    /// The host->client hint channel.
+    pub hint: FileChannel,
+    /// The host->client preimage channel.
+    pub preimage: FileChannel,
+}
+
+/// The orchestrator is responsible for starting the host and client program, and managing the
+/// communication between them. It is the entry point for the host to run a given module.
+///
+/// This trait is specific to running both the host and client program in-process. For detached
+/// mode, see [DetachedHostOrchestrator].
+#[async_trait]
+pub trait HostOrchestrator {
+    /// A collection of the providers that the host can use to reference remote resources.
+    type Providers;
+
+    /// Instantiates the providers for the host's fetcher.
+    async fn create_providers(&self) -> Result<Option<Self::Providers>>;
+
+    /// Constructs the [KeyValueStore] for the host.
+    ///
+    /// [KeyValueStore]: crate::KeyValueStore
+    fn create_key_value_store(&self) -> Result<SharedKeyValueStore>;
+
+    /// Creates a [Fetcher] for the host program's preimage server.
+    fn create_fetcher(
+        &self,
+        providers: Option<Self::Providers>,
+        kv_store: SharedKeyValueStore,
+    ) -> Option<Arc<RwLock<impl Fetcher + Send + Sync + 'static>>>;
+
+    /// Runs the client program natively and returns the exit code.
+    async fn run_client_native(
+        hint_reader: HintWriter<NativeChannel>,
+        oracle_reader: OracleReader<NativeChannel>,
+    ) -> Result<()>;
+
+    /// Starts the host and client program in-process.
+    async fn start(&self) -> Result<()> {
```

Comment:
This doesn't really just "start" the host and client though - it persistently runs them